[Ready and tested and microwaved] Rework the battle arcade game in the honor of fallen gamers

Caldony opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Reworks the battle arcade game in a more interesting, engaging vidya game.
The player will have 4 actions to choose from :

- light attack : just a normal light attack, not that much damage as is.

- defend : gain back some MP and reduce the damage of non magical attack.

- counter attack : if the enemy attacks you next turn with a non magical attack you will deal twice the amount of damage you took to the enemy. Costs 10MP

- power attack : deals twice the amount of damage compared to a light attack.
costs 20MP

and the enemy will have 3 of 6 randomly chosen passive which will dictate its actions

- smart : won't fall for your counter attacks ever but is heavily reliant on other passives.
weakpoint: defend,defend,light attack. Confusion as to why you would ever save up on MP only to use an MP free attack

- magical: constantly regains MP and will often blast you up for a fuckton of damage, also the only passive that doesn't have a weakpoint related to combos.
weakpoint: spam defend until your MP are equal or above 50. Way to flex on the magician

- short temper: short temper will always fall for a counter attack unless they are smart, otherwise they will charge at you for a lot of damage hurting themselves in the process or missing entirely.
weakpoint: (counter attack,counter attack,counter attack). spamming the same bullshit move is the best way of pissing off someone.

- poisonous: will deal 3 damage at the end of their turn no matter what. Pretty straight forward.
weakpoint: light attack light attack, light attack. use your weapon as a fan to make the fucker choke on their own poison

- shotgun: will augment the damage of nearly any attacks used, because the shotgun never miss
weakpoint : defend,defend,power attack. take the guy by surprise with a shove and shoot him with his own shotgun.

- chonker: gives more HP, one of the weakest passive if not for the fact that its weakpoint is expensive to trigger.
weakpoint : power attack,power attack, power attack. Tip them over, cost 60MP which is like 6 turns of defending. You should probably deal with other passive first.

ho right weakpoints. while the three passives may seem overwhelming at first, you CAN disable each passive by triggering their weakpoint, when all weakpoints are triggered, you will be able to do 10 time as much damage next turn, basically finishing most fights in a single bonk. most weakpoints require a certain combo of attack with the exception of the magical passive.

it's all about paying attention to the fight and actually be willing to experiment. It sounds complicated, but really the game is more about figuring the enemy's passive and choosing which one to take out first than just "unga me hit man until man give toy"

Do note this does NOT (yet) apply to the upgraded computer version of the game, which is kind of ironic because I kind of thought of making this PR when I was bored through space playing the vidya game on my space computer

the emag mode gives cuban pete ALL the passive among a special countdown, more HP and giving the player slightly more HP too. When the counter reaches 0 the player will be insta killed, the countdown has been optimised so that you will need to know the exact gaming tactic and a bunch of luck to succeed, the optimised tactic is :
defend 3 times to deactivate the magic passive, use counter attack 3 times to deactive short temper, then use defend two times in a row followed by 3 light attacks to disable both smart and poisonous. Then Finish that off with two defend and 3 power attacks to disable the shotgun and chonker passive.

if you SOMEHOW survived until then, hit him right into his weakpoint and finish him before the counter reaches 0.

in other word, you need to be the ultimate gamer to outgame cuban pete. Or you know, get lucky as fuck.

TODO: 

- [x] give a proper difficulty setting for emags.

- [x] give the G A M I N G skill more things in the game than just more damage.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
The battle arcade game sucked, there's no way around it. Like holy shit, boss that heals already aren't too much fun, but when literally every single one of your actions are random as hell and you basically have to roll the dice until you get a few lucky turns, it isn't fun, it's tedious.

This update is trying to make the game more interesting, and actually reward G A M E R S for trying to figure out a game instead of just spamming the same button over and over again until they win.

also my version use timers and not sleep so you know there's that.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: NT finally decided to update the arcade battle video game after its terrible space game reviews from syndicate gaming.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

caluany committed:
reworks the arcade game completly

gaming

caluany committed:
parenthesis more like orphanesis

MrDoomBringer commented:
pog pr pog pr
really this is great

ShibaInuLord commented:
this is actually pog holy shit

caluany committed:
adds the "emag mode"

emag mode is simple, cuban pete has all 6 traits at his disposal. The current best tactic I could find is 4 defend in a row to get rid of the magical trait, followed by 3 power attacks, which will both get rid of the shotgun and chonker trait

then hope to fucking hell that the short temper trait keeps cuban pete from hitting you and spam light attacks until he hopefully dies. Counter attacking is basically useless because of the smart/magical combo so don't even try to do that.

caluany committed:
fixes a fucky wucky

fucked up a thing and you weren't able to restart a game, also rebalance a few things.

caluany committed:
replace nearly every list by a lazylist

caluany committed:
fixes a bug with the arcade action

the arcade action didn't properly translate the href so I just made the enemy_action take a string directly instead of the entire href_list

caluany committed:
gives cuban pete a special bomb move

cuban pete works in a timer, after 18 turn the bomb will explode and basically reduce the player to fucking dust. It took a bunch of testing but I did manage to balance it relatively well, better gaming skill will also give you more HP

caluany committed:
makes the finishing move last through multiple turns

Now when the finishing move is ready you'll be able to use it whenever you want not just next turn, even multiple time in a row, unless you restart a game of course.

caluany committed:
makes the shotgun passive change the chosen_weapon

also slightly reform a few things so that the "bomb" text is always before any passive.

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -149,18 +150,77 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+	while(LAZYLEN(current_enemy_passive) < max_passive)
+		var/picked_passive = pick(passive_available)
+		LAZYADD(current_enemy_passive, picked_passive)
+		passive_available -= picked_passive
```
Use pickntake here.

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -149,18 +150,77 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+	current_enemy_passive = null
+
+	var/list/passive_available = enemy_passive_list.Copy()
+	while(LAZYLEN(current_enemy_passive) < max_passive)
```
Just make this a for loop.
```
for(var/i in 1 to max_passive)
```

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +249,297 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
-	popup.set_content(dat)
-	popup.set_title_image(user.browse_rsc_icon(icon, icon_state))
-	popup.open()
+	if(user.mind?.key) //mainly here to avoid a runtime when the player gets gibbed when losing the emag mode.
```
Check for user.client instead.

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +249,297 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
 
-/obj/machinery/computer/arcade/battle/proc/arcade_action(mob/user)
+
+/obj/machinery/computer/arcade/battle/proc/arcade_action(mob/user,player_stance,attackamt)
```
Document this proc.

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +249,297 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+		if(bomb_cooldown == 6)
+			temp += "<br><center><h3>[enemy_name]'s hand brushes the remote linked to the bomb, your heart skipped a beat. <center><h3>"
+		if(bomb_cooldown == 2)
+			temp += "<br><center><h3>[enemy_name] is going to press the button! it's now or never! <center><h3>"
```
Capitalize ``It's``

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +249,297 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+			player_hp -= attack_amount * 1000 //hey it's a maxcap we might as well go all in
+		bomb_cooldown--
+
+//yeah I used the shotgun as a passive, you know why? because the shotgun gives +5 attack which is pretty good
```
This comment indentation is extremely cursed.
Please tab them in by one.

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +249,297 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+			var/added_temp
+			var success = pick(TRUE,FALSE)
+
+			if(success)
```
just do
```
if(rand())
```

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +249,297 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+	if("magical" in current_enemy_passive)
+		if(player_mp >= 50)
+			temp += "<br><center><h3>the huge amount of magical energy you have acumulated throws [enemy_name] off balance!<center><h3>"
+			enemy_mp -= enemy_mp
```
```
enemy_mp = 0
```

caluany committed:
Nemvar how dare you help me improve my code

fixes a couple of thingies

caluany committed:
I like my comments like I like my women

not with an indentation in their skull

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +248,297 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+		else if(enemy_hp < 20 && enemy_mp >= 20 && enemy_stance == "defensive") //it's a pretty expensive spell so they can't spam it that much
+			temp += "<br><center><h3>[enemy_name] heal themselves with magic and gain back 20 hp!<center><h3>"
+			enemy_hp += 20
+			enemy_mp -= 30
```
This will cause negative mp.

Caldony replied:
ho right I rebalanced the MP cost but forgot to rebalance the MP check, good catch

nemvar commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +248,297 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+			player_hp -= 10
+			enemy_mp -= 20
+
+		else if(enemy_hp >= 20 && enemy_mp >= 30 && enemy_stance == "defensive")
```
This will never be true. The maximum value enemy_mp can have at this point is 24.

caluany committed:
fixes a few problem with the magical passive

caluany committed:
change gameover_check a little

make it so the player HP will be set to 1 if its under 0 and the enemy HP is under 0 for the rare case where the enemy kills you and die in their own turn.

Mainly here so it doesn't look weird that both you and your enemy are dead and yet you still win.

caluany committed:
put an anti runtime mesure

shouldn't happen but gives a default weapon of "null gun" if the weapons list is somehow empty. just in case.

Rohesie commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +250,301 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
-	if ((enemy_mp <= 0) || (enemy_hp <= 0))
+	if(enemy_hp <= 0)
 		if(!gameover)
+			if(timer_id)
+				deltimer(timer_id)
+			if(player_hp <= 0)
```
```suggestion
			if(timer_id)
				deltimer(timer_id)
				timer_id = null
			if(player_hp <= 0)
```

Rohesie commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -305,77 +560,64 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
-
-	if ((player_mp <= 0) || (player_hp <= 0))
+	else if(player_hp <= 0)
+		if(timer_id)
+			deltimer(timer_id)
```
```suggestion
		if(timer_id)
			deltimer(timer_id)
			timer_id = null
```

Rohesie commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +250,301 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+/obj/machinery/computer/arcade/battle/proc/enemy_action(player_stance,mob/user)
+	temp = ""
+
+	if(LAZYLEN(last_three_move) < 3) //we keep the last three action of the player in a list here
+		LAZYADD(last_three_move, player_stance)
+
+	else if(LAZYLEN(last_three_move) == 3)
+		LAZYREMOVE(last_three_move, last_three_move[1])
+		LAZYADD(last_three_move, player_stance)
+
+	else if(LAZYLEN(last_three_move) > 3)
+		last_three_move = null //this shouldn't even happen but we empty the list if it somehow goes above 3
```
```suggestion
	switch(LAZYLEN(last_three_move))
		if(0 to 2) //we keep the last three action of the player in a list here
			LAZYADD(last_three_move, player_stance)

		if(3)
			LAZYREMOVE(last_three_move, last_three_move[1])
			LAZYADD(last_three_move, player_stance)

		if(4 to INFINITY)
			last_three_move = null //this shouldn't even happen but we empty the list if it somehow goes above 3
```

Rohesie commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +250,301 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+
+	//if emagged, cuban pete will set up a bomb acting up as a timer. when it reaches 0 the player fucking dies
+	if(obj_flags & EMAGGED)
+		if(bomb_cooldown == 18)
+			temp += "<br><center><h3>[enemy_name] takes two valve tank and links them together, what's he planning?<center><h3>"
+		if(bomb_cooldown == 15)
+			temp += "<br><center><h3>[enemy_name] adds a remote control to the tan- ho god is that a bomb?<center><h3>"
+		if(bomb_cooldown == 12)
+			temp += "<br><center><h3>[enemy_name] throws the bomb next to you, you'r too scared to pick it up. <center><h3>"
+		if(bomb_cooldown == 6)
+			temp += "<br><center><h3>[enemy_name]'s hand brushes the remote linked to the bomb, your heart skipped a beat. <center><h3>"
+		if(bomb_cooldown == 2)
+			temp += "<br><center><h3>[enemy_name] is going to press the button! It's now or never! <center><h3>"
+		if(bomb_cooldown == 0)
+			player_hp -= attack_amount * 1000 //hey it's a maxcap we might as well go all in
+		bomb_cooldown--
```
```suggestion
		switch(bomb_cooldown)
			if(18)
				temp += "<br><center><h3>[enemy_name] takes two valve tank and links them together, what's he planning?<center><h3>"
			if(15)
				temp += "<br><center><h3>[enemy_name] adds a remote control to the tan- ho god is that a bomb?<center><h3>"
			if(12)
				temp += "<br><center><h3>[enemy_name] throws the bomb next to you, you'r too scared to pick it up. <center><h3>"
			if(6)
				temp += "<br><center><h3>[enemy_name]'s hand brushes the remote linked to the bomb, your heart skipped a beat. <center><h3>"
			if(2)
				temp += "<br><center><h3>[enemy_name] is going to press the button! It's now or never! <center><h3>"
			if(0)
				player_hp -= attack_amount * 1000 //hey it's a maxcap we might as well go all in
		bomb_cooldown--
```

Rohesie commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +250,301 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+		bomb_cooldown--
+
+	//yeah I used the shotgun as a passive, you know why? because the shotgun gives +5 attack which is pretty good
+	if("shotgun" in current_enemy_passive)
```
Try using associative lists, so you can check with `current_enemy_passive["shotgun"]` or with `LAZYACCESS(current_enemy_passive, "shotgun")`

Rohesie commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +250,301 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+			temp += "<br><center><h3>[enemy_name] is getting frustrated at all your counter attacks and throws a tantrum!<center><h3>"
+			enemy_hp -= attack_amount
+
+		else if(player_stance == "counter_attack" && enemy_hp > 30 && !("smart" in current_enemy_passive))
+			temp += "<br><center><h3>[enemy_name] took the bait and allowed you to counter attack for [attack_amount * 2] damage!<center><h3>"
+			player_hp -= attack_amount
+			enemy_hp -= attack_amount * 2
+			enemy_stance = "attack"
+
+		else if(player_stance == "counter_attack" && enemy_hp <= 30)
+			temp += "<br><center><h3>[enemy_name] is getting tired of your tricks and breaks through your counter with their [chosen_weapon]!<center><h3>"
+			player_hp -= attack_amount
+			enemy_stance = "attack"
```
```suggestion
		else if(player_stance == "counter_attack" && enemy_hp > 30)

			if(!("smart" in current_enemy_passive))
				temp += "<br><center><h3>[enemy_name] took the bait and allowed you to counter attack for [attack_amount * 2] damage!<center><h3>"
				player_hp -= attack_amount
				enemy_hp -= attack_amount * 2
				enemy_stance = "attack"

			else
				temp += "<br><center><h3>[enemy_name] is getting tired of your tricks and breaks through your counter with their [chosen_weapon]!<center><h3>"
				player_hp -= attack_amount
				enemy_stance = "attack"
```

Rohesie commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +250,301 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+	//in the case none of the previous passive triggered, Mainly here to set an enemy stance for passives that needs it like the magical passive.
+	if(!enemy_stance)
+		enemy_stance = pick("attack","defensive")
+		if(enemy_stance == "attack")
+			player_hp -= attack_amount
+			temp += "<br><center><h3>[enemy_name] attacks you for [attack_amount] points of damage with their [chosen_weapon]<center><h3>"
+			if(player_stance == "counter_attack")
+				enemy_hp -= attack_amount * 2
+				temp += "<br><center><h3>You counter [enemy_name]'s attack and deal [attack_amount * 2] points of damage!<center><h3>"
+
+		if(enemy_stance == "defensive" && enemy_mp < 15)
+			temp += "<br><center><h3>[enemy_name] take some time to get some mp back!<center><h3> "
+			enemy_mp += attack_amount
+
+		else if (enemy_stance == "defensive" && enemy_mp >= 15 && !("magical" in current_enemy_passive))
```
```suggestion
		if(enemy_stance == "attack")
			player_hp -= attack_amount
			temp += "<br><center><h3>[enemy_name] attacks you for [attack_amount] points of damage with their [chosen_weapon]<center><h3>"
			if(player_stance == "counter_attack")
				enemy_hp -= attack_amount * 2
				temp += "<br><center><h3>You counter [enemy_name]'s attack and deal [attack_amount * 2] points of damage!<center><h3>"

		else
			if(enemy_mp < 15)
				temp += "<br><center><h3>[enemy_name] take some time to get some mp back!<center><h3> "
				enemy_mp += attack_amount

			else if (!("magical" in current_enemy_passive))
```

caluany committed:
assoc list change

changes a few thing so that we can now use assoc list. along with changing a few if to switch per Rohesie's demand

Rohesie commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -189,107 +247,301 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
+
+///the enemy turn, the enemy's action entirely depend on their current passive and a teensy tiny bit of randomness
+/obj/machinery/computer/arcade/battle/proc/enemy_action(player_stance,mob/user)
+	temp = ""
```
Store this info as a list and at the end of it do `list.Join()`, possibly just `temp = list_temp.Join()`
It's much faster to do list operations than string ones here.

Rohesie commented:
Is this tested and working fine locally?

Caldony commented:
> 
> 
> Is this tested and working fine locally?

the title isn't just for show, but yes I did test everything multiple time, with no runtimes. Although I may have to add I tested it on the version of the game that's like a week out of date, but unless arcades have been changed it shouln't have too much of an impact

caluany committed:
change the way temp is modified during the enemy_setup() proc

basically as per Rohesie's request (again) there will be a temporary list that will add whatever it needs during the enemy setup then .Join() the list at the end.

caluany committed:
buff the finishing move

buff the finishing move so that it kills the enemy more reliably
